tip-distributor/src/claim_mev_workflow.rs

```diff
@@ -30,123 +36,408 @@ pub enum ClaimMevError {
 
     #[error(transparent)]
     JsonError(#[from] serde_json::Error),
+
+    #[error(transparent)]
+    AnchorError(anchor_lang::error::Error),
+
+    #[error("TDA not found for pubkey: {0:?}")]
+    TDANotFound(Pubkey),
+
+    #[error("Claim Status not found for pubkey: {0:?}")]
+    ClaimStatusNotFound(Pubkey),
+
+    #[error("Claimant not found for pubkey: {0:?}")]
+    ClaimantNotFound(Pubkey),
+
+    #[error(transparent)]
+    MaxFetchRetriesExceeded(#[from] solana_rpc_client_api::client_error::Error),
+
+    #[error("Failed after {attempts} retries. {remaining_transaction_count} remaining mev claim transactions, {failed_transaction_count} failed requests.",)]
+    MaxSendTransactionRetriesExceeded {
+        attempts: u64,
+        remaining_transaction_count: usize,
+        failed_transaction_count: usize,
+    },
+
+    #[error("Expected to have at least {desired_balance} lamports in {payer:?}. Current balance is {start_balance} lamports. Deposit {sol_to_deposit} SOL to continue.")]
+    InsufficientBalance {
+        desired_balance: u64,
+        payer: Pubkey,
+        start_balance: u64,
+        sol_to_deposit: u64,
+    },
 }
 
-pub fn claim_mev_tips(
-    merkle_root_path: &PathBuf,
-    rpc_url: &str,
+pub async fn claim_mev_tips(
+    merkle_trees: GeneratedMerkleTreeCollection,
+    rpc_url: String,
+    rpc_send_connection_count: u64,
+    max_concurrent_rpc_get_reqs: usize,
     tip_distribution_program_id: &Pubkey,
-    keypair_path: &PathBuf,
+    keypair: Arc<Keypair>,
+    max_loop_retries: u64,
+    max_loop_duration: Duration,
 ) -> Result<(), ClaimMevError> {
-    const MAX_RETRY_DURATION: Duration = Duration::from_secs(600);
+    let payer_pubkey = keypair.pubkey();
+    let blockhash_rpc_client = Arc::new(RpcClient::new_with_commitment(
+        rpc_url.clone(),
+        CommitmentConfig::finalized(),
+    ));
+    let rpc_clients = Arc::new(
+        (0..rpc_send_connection_count)
+            .map(|_| {
+                Arc::new(RpcClient::new_with_commitment(
+                    rpc_url.clone(),
+                    CommitmentConfig::confirmed(),
+                ))
+            })
+            .collect_vec(),
+    );
 
-    let merkle_trees: GeneratedMerkleTreeCollection =
-        read_json_from_file(merkle_root_path).expect("read GeneratedMerkleTreeCollection");
-    let keypair = read_keypair_file(keypair_path).expect("read keypair file");
+    let tree_nodes = merkle_trees
+        .generated_merkle_trees
+        .iter()
+        .flat_map(|tree| &tree.tree_nodes)
+        .collect_vec();
+    let minimum_rent = blockhash_rpc_client
```

Comment:
can you rename so it's obvious what type of account this is calculating min rent for? for example is it stake account, claim status? also is 0 data len correct?

Reply:
stake account was used previously to find the min rent. reading thru the code, the transfer should be at least the rent minimum for an regular wallet account rather than a claimstatus account, since we reclaim the claimstatus account after the epoch